### DISCUSSION TITLE
Subscribe to private Mercure updates

### DISCUSSION THREAD

**Discussion by janklan** [symfony/symfony#47718] [Q&A]
I'm scratching my head about what's the intended way to subscribe to private updates of a Mercure topic.

Following [ux-turbo docs](https://symfony.com/bundles/ux-turbo/current/index.html), using `{{ turbo_stream_listen("topicName") }}` works well for **public** updates, but private updates are not received, because the authorisation cookie is missing. 

[Symfony Mercure docs](https://symfony.com/doc/current/mercure.html#subscribing) talk about a twig function `mercure()`, which doesn't seem to be relevant when using the "ux-turbo way", `turbo_stream_listen`.

I know I can [generate the cookie programmatically in my controller](https://symfony.com/doc/current/mercure.html#programmatically-setting-the-cookie), but that's coupling the Twig template too tightly with said controller: what if I want to re-use a Twig snippet subscribing to a topic across the entire codebase?

I ended up generating the cookie calling `mercure()` and discarding its output, and while it works, it's extra-ugly:

```twig
{% do mercure('topicName', {'subscribe': ['topicName']}) %}
<div {{ turbo_stream_listen("topicName") }}>
...
</div>
```

I thought `turbo_stream_listen` would automatically generate the authorisation cookie with all the JWT claims necessary?

What am I missing? 

Cheers!

**Comment by Lacoi** [comment]
i have the same problem. Is there any solution or better workaround?

**Comment by gpetraroli** [comment]
@janklan did you find how to? I'm stuck on the same problem. Thx

**Reply by janklan** [reply]
I wrote a middleware that keeps tracks of topics I want to include in the JWT for more than once (entire session, certain amount of time, FIFO stack of up-to-N topics with the same prefix etc) and I'm simply generating tokens with those persistent topics in it.

The downside is that I expect this to break with too many topics included, and the token length obviously grows, but so far it seems to be well within what I consider OK and It works well.

**Comment by virgil422181** [comment]
Hi, 

This twig function is rendered by renderTurboStreamListen(Environment $env, $topic /* array $eventSourceOptions = [] */): string and is expected to pass an array on $eventSourceOptions  with a key "withCredentials". 

So if you execute with {{ turbo_stream_listen("topicName", null, {'withCredentials': true}) }} will work.
